smarts/core/opendrive_road_network.py

```diff
@@ -0,0 +1,1810 @@
+                    if in_lane.road.road_id != road.road_id:
+                        in_roads.add(in_lane.road)
+                for out_lane in lane.outgoing_lanes:
+                    if out_lane.road.road_id != road.road_id:
+                        out_roads.add(out_lane.road)
+            road.incoming_roads.extend(list(in_roads))
+            road.outgoing_roads.extend(list(out_roads))
+
+            for lane in road.lanes:
+                # Compute lane foes
+                result = [
+                    incoming
+                    for outgoing in lane.outgoing_lanes
+                    for incoming in outgoing.incoming_lanes
+                    if incoming != lane
+                ]
+                if lane.in_junction:
+                    in_roads = set(il.road for il in lane.incoming_lanes)
+                    for foe in lane.road.lanes:
+                        foe_in_roads = set(il.road for il in foe.incoming_lanes)
+                        if not bool(in_roads & foe_in_roads):
+                            result.append(foe)
+                lane.foes = list(set(result))
+
+            # recompute lane to left using road geometry if the map was converted from SUMO to OpenDRIVE
+            curr_leftmost_lane = road.lane_at_index(road.total_lanes - 1)
+            if curr_leftmost_lane and curr_leftmost_lane.lane_to_left[0] is None:
+                for other_road_id in self._roads:
+                    if other_road_id == road.road_id:
+                        continue
+                    other_road = self._roads[other_road_id]
+                    other_leftmost_lane = other_road.lane_at_index(
+                        other_road.total_lanes - 1
+                    )
+                    if other_leftmost_lane.lane_to_left[0] is not None:
+                        continue
+                    curr_leftmost_edge_shape, _ = road._shape(0)
+                    other_leftmost_edge_shape, _ = other_road._shape(0)
+                    edge_border_i = np.array(
+                        [curr_leftmost_edge_shape[0], curr_leftmost_edge_shape[-1]]
+                    )  # start and end position
+                    edge_border_j = np.array(
+                        [
+                            other_leftmost_edge_shape[-1],
+                            other_leftmost_edge_shape[0],
+                        ]
+                    )  # start and end position with reverse traffic direction
+
+                    # The edge borders of two lanes do not always overlap perfectly,
+                    # thus relax the tolerance threshold to 1
+                    if np.linalg.norm(edge_border_i - edge_border_j) < 1:
+                        curr_leftmost_lane.lane_to_left = other_leftmost_lane, False
+                        other_leftmost_lane.lane_to_left = curr_leftmost_lane, False
+
+        end = time.time()
+        elapsed = round((end - start) * 1000.0, 3)
+        self._log.info(f"Third pass: {elapsed} ms")
```

Comment:
I'm curious how long each pass is taking on average relative to the map size.  For example, if a user tries to use a huge map, how will the startup time scale?  Maybe it might be worth making a few simple graphs as to the time each pass takes relative to the number of roads (and/or number of "connections", or some other suitable measure of map complexity).

If it doesn't scale well, we might be forced to do this more like SumoRoadNetwork.

Reply:
We tried to load smaller and larger maps and observed that it  scaled really well. But @saulfield should comment further than this since he primarily looked into runtime issues with this.

Reply:
I'm going to write a few notes on performance based on the above discussion. Here are some numbers for a small map and a large map for comparison, run on my slow laptop:

```
build-scenario scenarios/od_merge/
INFO:OpenDriveRoadNetwork:Parsing .xodr file: 4.812 ms
INFO:OpenDriveRoadNetwork:First pass: 0.448 ms
INFO:OpenDriveRoadNetwork:Second pass: 256.43 ms
INFO:OpenDriveRoadNetwork:Third pass: 0.174 ms
```

```
build-scenario scenarios/od_newmarket/
INFO:OpenDriveRoadNetwork:Parsing .xodr file: 310.146 ms
INFO:OpenDriveRoadNetwork:First pass: 20.081 ms
INFO:OpenDriveRoadNetwork:Second pass: 5932.866 ms
INFO:OpenDriveRoadNetwork:Third pass: 1501.286 ms
```
As you can see, the 2nd and 3rd passes are the ones that are most impacted by map size. For the 2nd pass this is almost entirely taken up by converting the lane geometry to polygons. The `segment_size` parameter in the `LaneBoundary` class directly affects the runtime -- doubling the segment size will reduce the runtime by almost half. The 3rd pass is only expensive if it hits the branch where it calculates `lane_to_left` using road geometry.

Doubling the segment size and commenting out the `lane_to_left` branch results in the following for the same map:

```
build-scenario scenarios/od_newmarket/
INFO:OpenDriveRoadNetwork:Parsing .xodr file: 383.922 ms
INFO:OpenDriveRoadNetwork:First pass: 25.337 ms
INFO:OpenDriveRoadNetwork:Second pass: 3870.443 ms
INFO:OpenDriveRoadNetwork:Third pass: 4.427 ms
```

I personally think these numbers are reasonable for the fast and predictable runtime performance that comes from preloading the entire map at the beginning. However, I think there is quite a lot of performance still on the table. Some ideas:
- Make `segment_size` configurable or adaptive to cut down on geometry computation
- Parallelize lane geometry computation (they are all calculated independently so this would be pretty simple)
- Remove `lane_to_left` computations in 3rd pass (and make the map creator responsible for making those lanes part of the same road)